### PR TITLE
Add checkpoint to DEM solid objects

### PR DESCRIPTION
**Pull request by blaisb**
# Description of the problem

- The moving solid objects used in DEM simulations were not checkpointed at all. This is not good

# Description of the solution

- Add the full checkpointing of the solid objects
- Add a test to ensure that this works correctly

# How Has This Been Tested?

- A test has been added to check this feature. It works very well.

# Documentation

- I don't think anything needs to be documented in particular here?
-

### CODE REVIEW

File: source/dem/read_checkpoint.cc [hunk 1]
```diff
@@ -4,13 +4,15 @@ using namespace dealii;
...
 void
-read_checkpoint(TimerOutput                               &computing_timer,
-                const DEMSolverParameters<dim>            &parameters,
-                std::shared_ptr<SimulationControl>        &simulation_control,
-                PVDHandler                                &particles_pvdhandler,
-                PVDHandler                                &grid_pvdhandler,
-                parallel::distributed::Triangulation<dim> &triangulation,
-                Particles::ParticleHandler<dim>           &particle_handler)
+read_checkpoint(
+  TimerOutput                                             &computing_timer,
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/dem/read_checkpoint.cc:8]: That's the spacing the clang format auto indent is giving....

File: applications_tests/lethe-particles/restart_moving_receptacle.output [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+*****************************************************************
+Transient iteration: 20000    Time: 1        Time step: 5e-05   
+*****************************************************************
+| Variable                     | Min        | Max        | Average    | Total      | 
+| Contact list generation      | 0.0000e+00 | 1.1000e+01 | 1.7500e+00 | 7.0000e+01 | 
+| Velocity magnitude           | 2.1238e-06 | 9.7668e-02 | 1.1679e-03 | 1.1679e-01 | 
+| Angular velocity magnitude   | 3.6932e-04 | 7.2734e-01 | 5.3802e-02 | 5.3802e+00 | 
+| Translational kinetic energy | 3.0231e-15 | 6.3932e-06 | 6.5134e-08 | 6.5134e-06 | 
+| Rotational kinetic energy    | 1.8283e-15 | 7.0911e-09 | 1.1543e-10 | 1.1543e-08 |
```
Comment (OGaboriault) [inline applications_tests/lethe-particles/restart_moving_receptacle.output:130]: Is there a reason why you choose to use output log and not the subsection test?

File: source/core/serial_solid.cc [hunk 3]
```diff
@@ -484,77 +520,57 @@ SerialSolid<dim, spacedim>::write_output_results(
...
+  pvdhandler.save(prefix + "_solid_object_" + Utilities::int_to_string(id, 2));
 }
 
 template <int dim, int spacedim>
 void
-SerialSolid<dim, spacedim>::read_checkpoint(std::string /*prefix*/)
+SerialSolid<dim, spacedim>::read_checkpoint(std::string prefix)
 {
-  throw(std::runtime_error(
-    "Restarting DEM simulations with floating meshes is currently not possible."));
```
Comment (OGaboriault) [inline source/core/serial_solid.cc:510]: Maybe add a "### Fixed" in the Changelog ?

**Review by OGaboriault — APPROVED** [review]
Very nice! So existed to test the restart for my simulations! 

Minor comments... extra lines and Changelog mainly.

File: applications_tests/lethe-particles/restart_moving_receptacle.output (hunk 2 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-particles/restart_moving_receptacle.output:130]: Not really. I  guess I felt both were pretty much equivalent?

File: source/core/serial_solid.cc (hunk 3 above)
Comment (blaisb) [reply source/core/serial_solid.cc:510]: Will do so :)

**Pull request merged**